README.md
# d3-line-chunked

[![npm version](https://badge.fury.io/js/d3-line-chunked.svg)](https://badge.fury.io/js/d3-line-chunked)

A d3 plugin that renders a line with potential gaps in the data by styling the gaps differently from the defined areas. It also provides the ability to style arbitrary chunks of the defined data differently. Single points are rendered as circles and transitions are supported.

Blog: [Showing Missing Data in Line Charts](https://bocoup.com/weblog/showing-missing-data-in-line-charts)

Demo: http://peterbeshai.com/vis/d3-line-chunked/

![d3-line-chunked-demo](https://cloud.githubusercontent.com/assets/793847/18075172/806683f4-6e40-11e6-96bc-e0250adf0529.gif)

## Example Usage

```js
var lineChunked = d3.lineChunked()
  .x(function (d) { return xScale(d.x); })
  .y(function (d) { return yScale(d.y); })
  .curve(d3.curveLinear)
  .defined(function (d) { return d.y != null; })
  .lineStyles({
    stroke: '#0bb',
  });

d3.select('svg')
  .append('g')
    .datum(lineData)
    .transition()
    .duration(1000)
    .call(lineChunked);
```

### Example with multiple lines

```js
var lineChunked = d3.lineChunked()
  .x(function (d) { return xScale(d.x); })
  .y(function (d) { return yScale(d.y); })
  .defined(function (d) { return d.y != null; })
  .lineStyles({
    stroke: (d, i) => colorScale(i),
  });

var data = [
  [{ 'x': 0, 'y': 42 }, { 'x': 1, 'y': 76 }, { 'x': 2, 'y': 54 }],
  [{ 'x': 0, 'y': 386 }, { 'x': 1 }, { 'x': 2, 'y': 38 }, { 'x': 3, 'y': 192 }],
  [{ 'x': 0, 'y': 325 }, { 'x': 1, 'y': 132 }, { 'x': 2 }, { 'x': 3, 'y': 180 }]
];

// bind data
var binding = d3.select('svg').selectAll('g').data(data);

// append a `<g>` for each line
var entering = binding.enter().append('g');

// call lineChunked on enter + update
binding.merge(entering)
  .transition()
  .call(lineChunked);

// remove `<g>` when exiting
binding.exit().remove();
```

## Development

Get rollup watching for changes and rebuilding

```bash
npm run watch
```

Run a web server in the example directory

```bash
cd example
php -S localhost:8000
```

Go to http://localhost:8000


## Installing

If you use NPM, `npm install d3-line-chunked`. Otherwise, download the [latest release](https://github.com/pbeshai/d3-line-chunked/releases/latest).

Note that this project relies on the following d3 features and plugins:
- [d3-array](https://github.com/d3/d3-array)
- [d3-selection](https://github.com/d3/d3-selection)
- [d3-shape](https://github.com/d3/d3-interpolate)

If you are using transitions, you will also need:
- [d3-interpolate](https://github.com/d3/d3-interpolate)
- [d3-interpolate-path](https://github.com/pbeshai/d3-interpolate-path) (plugin)

The only thing not included in the default d3 v4 build is the plugin [d3-interpolate-path](https://github.com/pbeshai/d3-interpolate-path). You'll need to get that [separately](https://github.com/pbeshai/d3-interpolate-path#installing).

## API Reference

<a href="#lineChunked" name="lineChunked">#</a> d3.**lineChunked**()

Constructs a new generator for chunked lines with the default settings.


<a href="#_lineChunked" name="_lineChunked">#</a> *lineChunked*(*context*)

Render the chunked line to the given *context*, which may be either a [d3 selection](https://github.com/d3/d3-selection)
of SVG containers (either SVG or G elements) or a corresponding [d3 transition](https://github.com/d3/d3-transition). Reads the data for the line from the `datum` property on the  container.


<a href="#lineChunked_x" name="lineChunked_x">#</a> *lineChunked*.**x**([*x*])

Define an accessor for getting the `x` value for a data point. See [d3 line.x](https://github.com/d3/d3-shape#line_x) for details.


<a href="#lineChunked_y" name="lineChunked_y">#</a> *lineChunked*.**y**([*y*])

Define an accessor for getting the `y` value for a data point. See [d3 line.y](https://github.com/d3/d3-shape#line_y) for details.


<a href="#lineChunked_curve" name="lineChunked_curve">#</a> *lineChunked*.**curve**([*curve*])

Get or set the [d3 curve factory](https://github.com/d3/d3-shape#curves) for the line. See [d3 line.curve](https://github.com/d3/d3-shape#line_curve) for details.
Define an accessor for getting the `curve` value for a data point. See [d3 line.curve](https://github.com/d3/d3-shape#line_curve) for details.


<a href="#lineChunked_defined" name="lineChunked_defined">#</a> *lineChunked*.**defined**([*defined*])

Get or set *defined*, a function that given a data point (`d`) returns `true` if the data is defined for that point and `false` otherwise. This function is important for determining where gaps are in the data when your data includes points without data in them.

For example, if your data contains attributes `x` and `y`, but no `y` when there is no data available, you might set *defined* as follows:

```js
// sample data
var data = [{ x: 1, y: 10 }, { x: 2 }, { x: 3 }, { x: 4, y: 15 }, { x: 5, y: 12 }];

// returns true if d has a y value set
function defined(d) {
  return d.y != null;
}
```

It is only necessary to define this if your dataset includes entries for points without data.

The default returns `true` for all points.



<a href="#lineChunked_isNext" name="lineChunked_isNext">#</a> *lineChunked*.**isNext**([*isNext*])

Get or set *isNext*, a function to determine if a data point follows the previous. This function enables detecting gaps in the data when there is an unexpected jump.

For example, if your data contains attributes `x` and `y`, and does not include points with missing data, you might set **isNext** as follows:


```js
// sample data
var data = [{ x: 1, y: 10 }, { x: 4, y: 15 }, { x: 5, y: 12 }];

// returns true if current datum is 1 `x` ahead of previous datum
function isNext(previousDatum, currentDatum) {
  var expectedDelta = 1;
  return (currentDatum.x - previousDatum.x) === expectedDelta;
}
```

It is only necessary to define this if your data doesn't explicitly include gaps in it.

The default returns `true` for all points.


<a href="#lineChunked_transitionInitial" name="lineChunked_transitionInitial">#</a> *lineChunked*.**transitionInitial**([*transitionInitial*])

Get or set *transitionInitial*, a boolean flag that indicates whether to perform a transition on initial render or not. If true and the *context* that *lineChunked* is called in is a transition, then the line will animate its y value on initial render. If false, the line will appear rendered immediately with no animation on initial render. This does not affect any subsequent renders and their respective transitions.

The default value is `true`.

<a href="#lineChunked_extendEnds" name="lineChunked_extendEnds">#</a> *lineChunked*.**extendEnds**([*[xMin, xMax]*])

Get or set *extendEnds*, an array `[xMin, xMax]` specifying the minimum and maximum x pixel values
(e.g., `xScale.range()`). If defined, the undefined line will extend to the values provided,
otherwise it will end at the last defined points.


<a href="#lineChunked_accessData" name="lineChunked_accessData">#</a> *lineChunked*.**accessData**([*accessData*])

Get or set *accessData*, a function that specifies how to map from a dataset entry to the array of line data. This is only useful if your input data doesn't use the standard form of `[point1, point2, point3, ...]`. For example, if you pass in your data as `{ results: [point1, point2, point3, ...] }`, you would want to set accessData to `data => data.results`. For convenience, if your accessData function simply accesses a key of an object, you can pass it in directly: `accessData('results')` is equivalent to `accessData(data => data.results)`.

The default value is the identity function `data => data`.


<a href="#lineChunked_lineStyles" name="lineChunked_lineStyles">#</a> *lineChunked*.**lineStyles**([*lineStyles*])

Get or set *lineStyles*, an object mapping style keys to style values to be applied to both defined and undefined lines. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_styles).



<a href="#lineChunked_lineAttrs" name="lineChunked_lineAttrs">#</a> *lineChunked*.**lineAttrs**([*lineAttrs*])

Get or set *lineAttrs*, an object mapping attribute keys to attribute values to be applied to both defined and undefined lines. The passed in *lineAttrs* are merged with the defaults. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_attrs).

The default attrs are:

```js
{
  fill: 'none',
  stroke: '#222',
  'stroke-width': 1.5,
  'stroke-opacity': 1,
}
```



<a href="#lineChunked_gapStyles" name="lineChunked_gapStyles">#</a> *lineChunked*.**gapStyles**([*gapStyles*])

Get or set *gapStyles*, an object mapping style keys to style values to be applied only to undefined lines. It overrides values provided in *lineStyles*. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_styles).



<a href="#lineChunked_gapAttrs" name="lineChunked_gapAttrs">#</a> *lineChunked*.**gapAttrs**([*gapAttrs*])

Get or set *gapAttrs*, an object mapping attribute keys to attribute values to be applied only to undefined lines. It overrides values provided in *lineAttrs*. The passed in *gapAttrs* are merged with the defaults. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_attrs).

The default attrs are:

```js
{
  'stroke-dasharray': '2 2',
  'stroke-opacity': 0.2,
}
```


<a href="#lineChunked_pointStyles" name="lineChunked_pointStyles">#</a> *lineChunked*.**pointStyles**([*pointStyles*])

Get or set *pointStyles*, an object mapping style keys to style values to be applied to points. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_styles).



<a href="#lineChunked_pointAttrs" name="lineChunked_pointAttrs">#</a> *lineChunked*.**pointAttrs**([*pointAttrs*])

Get or set *pointAttrs*, an object mapping attr keys to attr values to be applied to points (circles). Note that if fill is not defined in *pointStyles* or *pointAttrs*, it will be read from the stroke color on the line itself. Uses syntax similar to [d3-selection-multi](https://github.com/d3/d3-selection-multi#selection_attrs).



<a href="#lineChunked_chunk" name="lineChunked_chunk">#</a> *lineChunked*.**chunk**([*chunk*])

Get or set *chunk*, a function that given a data point (`d`) returns the name of the chunk it belongs to. This is necessary if you want to have multiple styled chunks of the defined data. There are two reserved chunk names: `"line"` for the default line for defined data, and `"gap"` for undefined data. It is not recommended that you use `"gap"` in this function. The default value maps all data points to `"line"`.

For example, if you wanted all points with y values less than 10 to be in the `"below-threshold"` chunk, you could do the following:

```js
// sample data
var data = [{ x: 1, y: 5 }, { x: 2, y: 8 }, { x: 3, y: 12 }, { x: 4, y: 15 }, { x: 5, y: 6 }];

// inspects the y value to determine which chunk to use.
function chunk(d) {
  return d.y < 10 ? 'below-threshold' : 'line';
}
```


<a href="#lineChunked_chunkLineResolver" name="lineChunked_chunkLineResolver">#</a> *lineChunked*.**chunkLineResolver**([*chunkLineResolver*])

Get or set *chunkLineResolver*, a function that decides what chunk the line should be rendered in when given two adjacent defined points that may or may not be in the same chunk via `chunk()`. The function takes three parameters:

  * chunkNameLeft (*String*): The name of the chunk for the point on the left
  * chunkNameRight (*String*): The name of the chunk for the point on the right
  * chunkNames (*String[]*): The ordered list of chunk names from chunkDefinitions

It returns the name of the chunk that the line segment should be rendered in. By default it uses the order of the keys in the chunkDefinition object..

For example, if you wanted all lines between two different chunks to use the styling of the chunk that the left point belongs to, you could define *chunkLineResolver* as follows:

```js
// always take the chunk of the item on the left
function chunkLineResolver(chunkNameA, chunkNameB, chunkNames) {
  return chunkNameA;
}
```



CHANGES.md
# Change Log

## 1.4.0

- Added in support for arbitrarily styling different chunks of the defined data. This can be done through the new `chunk()`, `chunkDefinitions`, and `chunkLineResolver()` properties. See the README for details and the demo for examples.


## 1.3.0

- Increased gap line default opacity from 0.2 to 0.35 to make it more visible for lower contrast screens (#9)
- Getter/setters now check for arguments.length before deciding to get or set (#10)
- Circles will not disappear and reappear if the X position is the same. (#12)
- Circle animation now handles transitions with delays (#13)


example/example-gallery.js
/* eslint-disable */
/**
 * File to generate a path for a variety of different conditions
 */
(function (d3) {
  var exampleWidth = 300;
  var exampleHeight = 100;

  var x = d3.scaleLinear().domain([0, 10]).range([10, exampleWidth - 10]);
  var y = d3.scaleLinear().domain([0, 4]).range([exampleHeight - 10, 10]);

  var transitionDuration = 2500;
  var transitionDebug = false;

  var examples = [
    {
      label: 'Typical',
      render: function typicalExample(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, 1], [1, 2], [2, null], [3, null], [4, 1], [5, null], [6, 2], [7, 3], [8, null], [9, 1], [10, 1]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Typical with curve',
      render: function typicalExampleWithCurve(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .curve(d3.curveMonotoneX)
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, 1], [1, 2], [2, null], [3, null], [4, 1], [5, null], [6, 2], [7, 3], [8, null], [9, 1], [10, 1]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Many points, all defined',
      render: function manyPointsAllDefined(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, 1], [2, 2], [4, 1], [10, 0]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Undefined at ends',
      render: function undefinedAtEnds(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .lineStyles({ 'stroke-width': '10px' })
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, null], [1, null], [2, 1], [3, 3], [4, 2], [5, 2], [6, 0], [7, 1], [8, 1], [9, null], [10, null]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Undefined at ends + extendEnds',
      render: function undefinedAtEndsExtendEnds(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .extendEnds(x.range());

        var data = [[0, null], [1, null], [2, 1], [3, 3], [4, 2], [5, 2], [6, 0], [7, 1], [8, 1], [9, null], [10, null]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Lines along top and bottom edges',
      render: function topBottomEdges(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .lineStyles({ 'stroke-width': 4 });

        var data = [[0, 0], [4, 0], [5, 2], [6, 2], [7, 4], [10, 4]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Data length 1',
      render: function dataLength1(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, 1]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Empty Data',
      render: function emptyData(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'One undefined point',
      render: function oneDefinedPoint(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, null]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Many points, all undefined',
      render: function manyPointsUndefined(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, null], [1, null], [2, null], [3, null]];

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Typical with accessData',
      render: function typicalExample(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .accessData(data => data.results);

        var data = { results: [[0, 1], [1, 2], [2, null], [3, null], [4, 1], [5, null], [6, 2], [7, 3], [8, null], [9, 1], [10, 1]] };

        g.datum(data).call(chunked);
      },
    },
    {
      label: 'Transition: transitionInitial=true',
      transition: true,
      render: function transitionInitialTrue(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .debug(transitionDebug)
          .transitionInitial(true);

        var data = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];

        g.datum(data).transition().duration(transitionDuration).call(chunked);
      },
    },
    {
      label: 'Transition: full to missing',
      transition: true,
      render: function fullToMissing(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[0, 1], [2, 2], [4, 1], [6, 2], [8, 2], [9, 0]];
        var dataEnd = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: from point',
      transition: true,
      render: function fromPoint(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[5, 1]];
        var dataEnd = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: from point + extendEnds',
      transition: true,
      render: function fromPointExtendEnds(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .extendEnds(x.range())
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[5, 1]];
        var dataEnd = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: gap to line',
      transition: true,
      render: function gapToLine(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[0, 1], [2, 2], [4, 1], [5, null], [8, 2], [9, 0]];
        var dataEnd = [[0, 1], [2, 2], [4, 1], [6, 2], [8, 2], [9, 0]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: few -> many segments',
      transition: true,
      render: function fewToMany(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .isNext(function (prev, curr) { return curr[0] === prev[0] + 1; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[0, 1], [1, 2], [7, 0], [8, 1], [9, 0], [10, 1]];
        var dataEnd = [[0, 1], [1, 2], [3, 0], [4, 1], [6, 3], [7, 2], [9, 1], [10, 1]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: end segment overlap',
      transition: true,
      render: function endSegmentOverlap(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var x = d3
          .scaleLinear()
          .domain([3, 19])
          .range([10, exampleWidth - 10]);

        var y = d3
          .scaleLinear()
          .domain([2, 230])
          .range([exampleHeight - 10, 10]);

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d.x); })
          .y(function (d) { return y(d.y); })
          .defined(function (d) { return d.y != null; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [{"x":3,"y":13,"v":160},{"x":4,"y":230,"v":93},{"x":5,"v":149},{"x":6,"y":4,"v":207},{"x":7,"y":21,"v":96},{"x":8,"y":2,"v":128},{"x":9,"y":6,"v":151},{"x":10,"y":14,"v":224},{"x":11,"v":70},{"x":12,"y":36,"v":104},{"x":13,"y":9,"v":190},{"x":14,"v":202},{"x":15,"y":5,"v":67},{"x":16,"y":5,"v":177},{"x":17,"y":25,"v":79},{"x":18,"y":3,"v":201},{"x":19,"y":34,"v":125}];
        var dataEnd = [{"x":3,"y":63,"v":110},{"x":4,"y":16,"v":133},{"x":5,"y":45,"v":143},{"x":6,"y":3,"v":284},{"x":7,"y":6,"v":150},{"x":8,"y":22,"v":233},{"x":9,"v":207},{"x":10,"y":173,"v":109},{"x":11,"y":110,"v":80},{"x":12,"y":17,"v":133},{"x":13,"y":11,"v":192},{"x":14,"y":2,"v":131},{"x":15,"v":117},{"x":16,"y":149,"v":111},{"x":17,"y":99,"v":149},{"x":18,"y":20,"v":145},{"x":19,"y":10,"v":127}];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Multiple lines',
      transition: true,
      render: function endSegmentOverlap(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var x = d3
          .scaleLinear()
          .domain([3, 19])
          .range([10, exampleWidth - 10]);

        var y = d3
          .scaleLinear()
          .domain([2, 250])
          .range([exampleHeight - 10, 10]);

        var color = d3.scaleOrdinal(d3.schemeCategory10);

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d.x); })
          .y(function (d) { return y(d.y); })
          .defined(function (d) { return d.y != null; })
          .debug(transitionDebug)
          .lineStyles({
            stroke: (d, i) => color(i),
          })
          .transitionInitial(true);

        var dataStart = [[{"x":0,"y":42,"v":93},{"x":1,"y":7,"v":216},{"x":2,"y":5,"v":174},{"x":3,"y":5,"v":105},{"x":4,"y":12,"v":235},{"x":5,"y":108,"v":137},{"x":6,"v":36},{"x":7,"y":146,"v":122},{"x":8,"y":115,"v":223},{"x":9,"v":192},{"x":10,"y":145,"v":114},{"x":11,"y":21,"v":130},{"x":12,"v":64},{"x":13,"v":158},{"x":14,"y":58,"v":226},{"x":15,"y":7,"v":215},{"x":16,"y":44,"v":141},{"x":17,"y":5,"v":126},{"x":18,"y":39,"v":144},{"x":19,"y":28,"v":134}],
          [{"x":0,"y":38,"v":166},{"x":1,"y":11,"v":197},{"x":2,"y":38,"v":80},{"x":3,"y":19,"v":222},{"x":4,"v":140},{"x":5,"y":23,"v":100},{"x":6,"y":13,"v":161},{"x":7,"y":47,"v":152},{"x":8,"v":145},{"x":9,"v":143},{"x":10,"y":16,"v":51},{"x":11,"y":17,"v":180},{"x":12,"y":9,"v":148},{"x":13,"v":196},{"x":14,"y":24,"v":207},{"x":15,"y":2,"v":19},{"x":16,"y":4,"v":165},{"x":17,"v":77},{"x":18,"y":123,"v":108},{"x":19,"y":81,"v":234}],
          [{"x":0,"y":32,"v":155},{"x":1,"y":13,"v":192},{"x":2,"y":7,"v":157},{"x":3,"y":100,"v":176},{"x":4,"v":106},{"x":5,"y":10,"v":209},{"x":6,"y":26,"v":19},{"x":7,"v":109},{"x":8,"y":7,"v":247},{"x":9,"y":11,"v":172},{"x":10,"y":236,"v":115},{"x":11,"y":1,"v":91},{"x":12,"y":3,"v":180},{"x":13,"y":19,"v":195},{"x":14,"v":46},{"x":15,"y":3,"v":211},{"x":16,"v":183},{"x":17,"v":148},{"x":18,"y":60,"v":181},{"x":19,"y":10,"v":119}]];

        var dataEnd = [[{"x":0,"y":63,"v":276},{"x":1,"y":30,"v":230},{"x":2,"y":4,"v":139},{"x":3,"y":35,"v":93},{"x":4,"y":1,"v":265},{"x":5,"y":131,"v":206},{"x":6,"y":49,"v":65},{"x":7,"y":10,"v":183},{"x":8,"v":186},{"x":9,"y":36,"v":175},{"x":10,"y":31,"v":28},{"x":11,"y":2,"v":137},{"x":12,"y":15,"v":52},{"x":13,"y":8,"v":200},{"x":14,"y":8,"v":125},{"x":15,"y":79,"v":161},{"x":16,"y":55,"v":241},{"x":17,"y":1,"v":173},{"x":18,"y":6,"v":137},{"x":19,"y":27,"v":120}],
          [{"x":0,"y":5,"v":153},{"x":1,"y":36,"v":244},{"x":2,"y":43,"v":57},{"x":3,"y":15,"v":102},{"x":4,"y":281,"v":228},{"x":5,"y":16,"v":174},{"x":6,"y":41,"v":32},{"x":7,"y":45,"v":144},{"x":8,"v":115},{"x":9,"y":27,"v":99},{"x":10,"y":115,"v":190},{"x":11,"v":113},{"x":12,"y":26,"v":154},{"x":13,"y":26,"v":131},{"x":14,"v":211},{"x":15,"v":192},{"x":16,"y":48,"v":103},{"x":17,"y":4,"v":126},{"x":18,"y":6,"v":141},{"x":19,"y":23,"v":187}]];


        function updateChart(data) {
          var binding = g.selectAll('g').data(data);
          var entering = binding.enter().append('g');
          binding.merge(entering)
            .transition().duration(transitionDuration / 3)
            .call(chunked);
          binding.exit().remove();
        }

        updateChart(dataStart);

        setTimeout(function () {
          updateChart(dataEnd);
          setTimeout(function () {
            updateChart(dataStart);
          }, transitionDuration / 2);
        }, transitionDuration / 2);

      },
    },
    {
      label: 'Transition: null to null',
      render: function typicalExample(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; });

        var data = [[0, null]];

        g.datum(data).call(chunked).transition().call(chunked);
      },
    },
    {
      label: 'Transition: has point same X location',
      transition: true,
      render: function fromPoint(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];
        var dataEnd = [[0, 1], [2, 2], [4, 1], [5, null], [6, 0], [7, null], [8, 2], [9, 0]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },
    {
      label: 'Transition: transition has delay',
      transition: true,
      render: function fromPoint(root) {
        var g = root.append('svg')
          .attr('width', exampleWidth)
          .attr('height', exampleHeight)
          .append('g');

        var chunked = d3.lineChunked()
          .x(function (d) { return x(d[0]); })
          .y(function (d) { return y(d[1]); })
          .defined(function (d) { return d[1] !== null; })
          .isNext(function (prev, curr) { return curr[0] === prev[0] + 1; })
          .debug(transitionDebug)
          .transitionInitial(false);

        var dataStart = [[0, 1], [2, 2], [4, 1], [5, null], [6, 2], [7, null], [8, 2], [9, 0]];
        var dataEnd = [[0, 1], [1, 2], [3, 0], [4, 1], [6, 3], [7, 2], [9, 1], [10, 1]];

        g.datum(dataStart).call(chunked);
        setTimeout(function () {
          g.datum(dataEnd).transition().delay(2000).duration(transitionDuration).call(chunked);
        }, transitionDuration / 4);
      },
    },

  ];



  // render the gallery
  var galleryRoot = d3.select('.example-gallery');

  // append transition timing slider
  var transitionSlider = galleryRoot.append('div');
  transitionSlider.append('strong').text('Transition Duration')
  var transitionSliderValue = transitionSlider.append('span').text(transitionDuration)
    .style('margin-left', '10px');
  transitionSlider.append('input')
    .style('display', 'block')
    .attr('type', 'range')
    .attr('min', 0)
    .attr('max', 5000)
    .attr('value', transitionDuration)
    .on('change', function () {
      transitionDuration = parseFloat(this.value);
      transitionSliderValue.text(transitionDuration);
    });
  var transitionDebugControl = transitionSlider.append('label');
  transitionDebugControl.append('input')
    .attr('type', 'checkbox')
    .attr('checked', transitionDebug ? true : null)
    .on('change', function () {
      transitionDebug = this.checked;
    });
  transitionDebugControl.append('span')
    .text('Debug');


  examples.forEach(function (example) {
    var div = galleryRoot.append('div').attr('class', 'example');

    if (example.transition) {
      div.append('button')
        .style('float', 'right')
        .text('Play')
        .on('click', function () {
          div.select('svg').remove();
          example.render(div);
        });
    }

    div.append('h3')
      .text(example.label);



    example.render(div);
  });
})(window.d3);